File: 025. Reverse Nodes in k-Group/Reverse Nodes in k-Group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        dummy = ListNode(-1)
        dummy.next = head
        pre = dummy
        cur = pre
        num = 0
        while  cur.next: 
            num += 1
            cur = cur.next
        while num >= k:
            cur = pre.next
            for i in xrange(1,k):
                t = cur.next
                cur.next = t.next
                t.next = pre.next
                pre.next = t
            
            pre = cur
            num -= k
        
        return dummy.next
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head or k == 1: return head
        dummy = ListNode(-1)
        dummy.next = head
        pre = dummy
        cur = head
        
        i = 1
        while cur:
            if i % k == 0:
                pre = self.reverse(pre, cur.next)
                cur = pre.next
            else:
                cur = cur.next
            i += 1
        return dummy.next
    def reverse(self, pre, next):
        last = pre.next
        cur = last.next
        while cur != next:
            last.next = cur.next
            cur.next = pre.next
            pre.next = cur
            cur = last.next
        
        
        return last
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        cur = head
        for i in xrange(k):
            if not cur: return head
            cur = cur.next
        
        new_head = self.reverse(head, cur)
        head.next = self.reverseKGroup(cur, k)
        return new_head
    def reverse(self, head, tail):
       
        pre = tail
        while head != tail:
            t = head.next
            head.next = pre
            pre = head
            head = t
            
        
        return pre

File: 025. Reverse Nodes in k-Group/Reverse Nodes in k-Group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode cur = pre;
        int num = 0;
        while((cur = cur.next) != null) num++;
        while(num >= k){
            cur = pre.next;
            for(int i = 1; i < k; i++){
                ListNode t = cur.next;
                cur.next = t.next;
                t.next = pre.next;
                pre.next = t;
            }
            pre = cur;
            num -= k;
        }
        return dummy.next;
    }
}
**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head == null || k == 1) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        
        
        ListNode pre = dummy;
        ListNode cur = head;
        
       
        for(int i = 1; cur != null; i++){
            if(i % k == 0){
                pre = reverse(pre, cur.next);
                cur = pre.next;
            } else {
                cur = cur.next;
            }
        }
        
        
        return dummy.next;
    }
    ListNode reverse(ListNode pre, ListNode next) {
        ListNode last = pre.next;
        ListNode cur = last.next;
        while(cur != next){
            last.next = cur.next;
            cur.next = pre.next;
            pre.next = cur;
            cur = last.next;
        }
        
        return last;
    }
}
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode cur = head;
        for (int i = 0; i < k; ++i) {
            if (cur == null) return head;
            cur = cur.next;
        }
        ListNode new_head = reverse(head, cur);
        head.next = reverseKGroup(cur, k);
        return new_head;
    }
    public ListNode reverse(ListNode head, ListNode tail) {
       
        ListNode pre = tail;
        while(head != tail){
            ListNode t = head.next;
            head.next = pre;
            pre = head;
            head = t;
            
        }
        return pre;
    }
}
